Sources/SecureXPC/Codable Property Wrappers/File Descriptor XPC Wrappers.swift

```diff
@@ -0,0 +1,184 @@
+//
+//  File Descriptor XPC Wrappers.swift
+//  SecureXPC
+//
+//  Created by Josh Kaplan on 2022-07-10
+//
+
+import Foundation
+import System
+
+// MARK: common implementation
+
+private protocol FileDescriptorCodable: Codable {
+    init(descriptor: CInt, closeOnEncode: Bool)
+    var descriptor: CInt { get }
+    var closeOnEncode: Bool { get }
+    func close() throws
+}
+
+private enum CodingKeys: String, CodingKey {
+    case descriptor
+    case closeOnEncode
+}
+
+extension FileDescriptorCodable {
+    public func encode(to encoder: Encoder) throws {
+        let container = try XPCEncoderImpl.asXPCEncoderImpl(encoder).xpcContainer(keyedBy: CodingKeys.self)
+        guard let encodedDescriptor = xpc_fd_create(self.descriptor) else {
+            let context = EncodingError.Context(codingPath: container.codingPath,
+                                                debugDescription: "Encoding failed for \(self.descriptor)",
+                                                underlyingError: nil)
+            throw EncodingError.invalidValue(self.descriptor, context)
+        }
+        
+        if self.closeOnEncode {
+            try self.close()
+        }
+        container.encode(encodedDescriptor, forKey: CodingKeys.descriptor)
+        try container.encode(self.closeOnEncode, forKey: CodingKeys.closeOnEncode)
```

Comment:
This is a weird situation. We failed to close it, but we're going to try to encode it anyway. Not sure what the "safe" semantics should be here.

Reply:
If it fails to close, shouldn't that cause an error to be thrown and therefore the encoding won't happen?

Reply:
Oh right, true.

What about the result of `xpc_fd_create`? The docs say it acts like `dup()` was called on the parameter. If an error is thrown here, should we catch it, close the `encodedDescriptor`, then rethrow the error?

Reply:
My interpretation is it's _like_ `dup()`, but not actually the same in regard to the resulting `xpc_object_t` (or something opaquely referenced by it) needs to be explicitly closed. I say that because it appears to be impossible to close it. Once it's been encoded as an `xpc_object_t` the only way to get back a file descriptor is to call `xpc_fd_dup()` which itself says:

> Returns a file descriptor that is equivalent to the one boxed by the file file descriptor object.

and also says:

> Multiple invocations of xpc_fd_dup() will not return the same file descriptor number, but they will return descriptors that are equivalent, as though they had been created by dup(2).

My reading of it is an attempt to get a file descriptor using `xpc_fd_dup()` in order to close it would be pointless as it's `dup()`ing it each time one is requested - it's not a "true" unboxing of the original value.